Added Cosine Similarity metric

GiannisVagionakis opened the pull request:
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

## What does this PR do?
Resolves #75

I am co-worker of @N-georgakopoulos (issue assigned to him: https://github.com/PyTorchLightning/metrics/issues/75) and so I made this PR with good faith that it solves the issue sufficiently as regards the basic implementation, being more than happy to follow any advice or directive from someone more exprerienced

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

pep8speaks commented:
Hello @GiannisVagionakis! Thanks for opening this PR. 

* In the file [`torchmetrics/classification/__init__.py`](https://github.com/PyTorchLightning/metrics/blob/3e69f4c1eca4a06e180d214270eccd2b48ed0844/torchmetrics/classification/__init__.py):

> [Line 33:75](https://github.com/PyTorchLightning/metrics/blob/3e69f4c1eca4a06e180d214270eccd2b48ed0844/torchmetrics/classification/__init__.py#L33): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment

* In the file [`torchmetrics/classification/cosine_similarity.py`](https://github.com/PyTorchLightning/metrics/blob/3e69f4c1eca4a06e180d214270eccd2b48ed0844/torchmetrics/classification/cosine_similarity.py):

> [Line 20:121](https://github.com/PyTorchLightning/metrics/blob/3e69f4c1eca4a06e180d214270eccd2b48ed0844/torchmetrics/classification/cosine_similarity.py#L20): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (122 > 120 characters)
> [Line 22:1](https://github.com/PyTorchLightning/metrics/blob/3e69f4c1eca4a06e180d214270eccd2b48ed0844/torchmetrics/classification/cosine_similarity.py#L22): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 77:9](https://github.com/PyTorchLightning/metrics/blob/3e69f4c1eca4a06e180d214270eccd2b48ed0844/torchmetrics/classification/cosine_similarity.py#L77): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 81:62](https://github.com/PyTorchLightning/metrics/blob/3e69f4c1eca4a06e180d214270eccd2b48ed0844/torchmetrics/classification/cosine_similarity.py#L81): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 81:64](https://github.com/PyTorchLightning/metrics/blob/3e69f4c1eca4a06e180d214270eccd2b48ed0844/torchmetrics/classification/cosine_similarity.py#L81): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 98:84](https://github.com/PyTorchLightning/metrics/blob/3e69f4c1eca4a06e180d214270eccd2b48ed0844/torchmetrics/classification/cosine_similarity.py#L98): [W292](https://duckduckgo.com/?q=pep8%20W292) no newline at end of file

* In the file [`torchmetrics/functional/classification/__init__.py`](https://github.com/PyTorchLightning/metrics/blob/3e69f4c1eca4a06e180d214270eccd2b48ed0844/torchmetrics/functional/classification/__init__.py):

> [Line 31:87](https://github.com/PyTorchLightning/metrics/blob/3e69f4c1eca4a06e180d214270eccd2b48ed0844/torchmetrics/functional/classification/__init__.py#L31): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment

* In the file [`torchmetrics/functional/classification/cosine_similarity.py`](https://github.com/PyTorchLightning/metrics/blob/3e69f4c1eca4a06e180d214270eccd2b48ed0844/torchmetrics/functional/classification/cosine_similarity.py):

> [Line 22:1](https://github.com/PyTorchLightning/metrics/blob/3e69f4c1eca4a06e180d214270eccd2b48ed0844/torchmetrics/functional/classification/cosine_similarity.py#L22): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 31:1](https://github.com/PyTorchLightning/metrics/blob/3e69f4c1eca4a06e180d214270eccd2b48ed0844/torchmetrics/functional/classification/cosine_similarity.py#L31): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 34:14](https://github.com/PyTorchLightning/metrics/blob/3e69f4c1eca4a06e180d214270eccd2b48ed0844/torchmetrics/functional/classification/cosine_similarity.py#L34): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 34:16](https://github.com/PyTorchLightning/metrics/blob/3e69f4c1eca4a06e180d214270eccd2b48ed0844/torchmetrics/functional/classification/cosine_similarity.py#L34): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 39:29](https://github.com/PyTorchLightning/metrics/blob/3e69f4c1eca4a06e180d214270eccd2b48ed0844/torchmetrics/functional/classification/cosine_similarity.py#L39): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 39:41](https://github.com/PyTorchLightning/metrics/blob/3e69f4c1eca4a06e180d214270eccd2b48ed0844/torchmetrics/functional/classification/cosine_similarity.py#L39): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 44:1](https://github.com/PyTorchLightning/metrics/blob/3e69f4c1eca4a06e180d214270eccd2b48ed0844/torchmetrics/functional/classification/cosine_similarity.py#L44): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 44:63](https://github.com/PyTorchLightning/metrics/blob/3e69f4c1eca4a06e180d214270eccd2b48ed0844/torchmetrics/functional/classification/cosine_similarity.py#L44): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 44:65](https://github.com/PyTorchLightning/metrics/blob/3e69f4c1eca4a06e180d214270eccd2b48ed0844/torchmetrics/functional/classification/cosine_similarity.py#L44): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 68:1](https://github.com/PyTorchLightning/metrics/blob/3e69f4c1eca4a06e180d214270eccd2b48ed0844/torchmetrics/functional/classification/cosine_similarity.py#L68): [W391](https://duckduckgo.com/?q=pep8%20W391) blank line at end of file

Do see the [Hitchhiker's guide to code style](https://goo.gl/hqbW4r)

Borda reviewed (commented):
@GiannisVagionakis great additions, mind check
- [ ] add docs
- [ ] add tests

SkafteNicki commented on torchmetrics/classification/cosine_similarity.py:
```diff
@@ -0,0 +1,102 @@
+        self.add_state("total", default=tensor(0), dist_reduce_fx="sum")
+        self.reduction = reduction
+
+        if not 0 < threshold < 1:
+            raise ValueError("The `threshold` should lie in the (0,1) interval.")
+        self.threshold = threshold
```
```suggestion
```

SkafteNicki commented on torchmetrics/classification/cosine_similarity.py:
```diff
@@ -0,0 +1,102 @@
+
+    def __init__(
+        self,
+        threshold: float = 0.5,
```
```suggestion
```

SkafteNicki commented on torchmetrics/classification/cosine_similarity.py:
```diff
@@ -0,0 +1,102 @@
+        dist_sync_on_step: bool = False,
+        process_group: Optional[Any] = None,
+        dist_sync_fn: Callable = None,
+        reduction: str = 'sum'
```
let this be the first parameter

SkafteNicki commented on torchmetrics/classification/cosine_similarity.py:
```diff
@@ -0,0 +1,102 @@
+            raise ValueError("The `threshold` should lie in the (0,1) interval.")
+        self.threshold = threshold
+
+    def update(self, preds: Tensor, target: Tensor, reduction="sum"):
```
```suggestion
    def update(self, preds: Tensor, target: Tensor):
```

SkafteNicki commented on torchmetrics/classification/cosine_similarity.py:
```diff
@@ -0,0 +1,102 @@
+        Args:
+            preds: Predictions from model (probabilities, logits or labels)
+            target: Ground truth labels
+            reduction : how to reduce over the batch dimension using sum, mean or
+                        taking the individual scores
```
copy this text to the docstring of the init method
```suggestion

```

SkafteNicki commented on torchmetrics/classification/cosine_similarity.py:
```diff
@@ -0,0 +1,102 @@
+
+        self.correct += correct
+        self.total += total
+        self.reduction = reduction
```
```suggestion

```

SkafteNicki commented on torchmetrics/classification/cosine_similarity.py:
```diff
@@ -0,0 +1,102 @@
+               will be used to perform the all gather.
+           reduction: The method of reducing along the batch dimension using sum, mean or
+                        taking the individual scores
+       Raises:
+           ValueError:
+               If ``threshold`` is not between ``0`` and ``1``.
```
```suggestion
```

SkafteNicki commented on torchmetrics/classification/cosine_similarity.py:
```diff
@@ -0,0 +1,102 @@
+        between targets and predictions:
+       Accepts all input types listed in :ref:`references/modules:input types`.
+       Args:
+           threshold:
+               Threshold for transforming probability or logit predictions to binary (0,1) predictions, in the case
+               of binary or multi-label inputs. Default value of 0.5 corresponds to input being probabilities.
```
should be exchanged with a docstring of the `reduction` arg
```suggestion

```

SkafteNicki reviewed (commented):
Thanks for wanting to contribute. Most comments are about the `threshold` parameter which seems to be a copy-paste error from another metric :]

GiannisVagionakis commented:
Hi, writting some documentation now, with respect to updating the pull request by the weekend! I would like however to understand a little bit more about input types? Is the implementation example i've put in the comments correct?

Borda commented on torchmetrics/classification/cosine_similarity.py:
```diff
@@ -0,0 +1,88 @@
+       Computes the `Cosine Similarity <https://en.wikipedia.org/wiki/Cosine_similarity>`_
+        between targets and predictions:
+       Accepts all input types listed in :ref:`references/modules:input types`.
+       Args:
```
```suggestion

       Args:
```

Borda commented on torchmetrics/classification/cosine_similarity.py:
```diff
@@ -0,0 +1,88 @@
+           dist_sync_fn:
+               Callback that performs the allgather operation on the metric state. When ``None``, DDP
+               will be used to perform the all gather.
+       Example:
```
```suggestion

       Example:
```

Borda commented on torchmetrics/classification/cosine_similarity.py:
```diff
@@ -0,0 +1,88 @@
+        """
+        Update state with predictions and targets. See :ref:`references/modules:input types` for more information
+        on input types.
+        Args:
```
```suggestion

        Args:
```

Borda commented on torchmetrics/functional/classification/cosine_similarity.py:
```diff
@@ -0,0 +1,62 @@
+        Computes the `Cosine Similarity <https://en.wikipedia.org/wiki/Cosine_similarity>`_
+        between targets and predictions:
+        Accepts all input types listed in :ref:`references/modules:input types`.
+        Args:
```
```suggestion

        Args:
```

Borda commented:
> Hi, writting some documentation now, with respect to updating the pull request by the weekend!

if so we may include it in net release :]

> I would like however to understand a little bit more about input types? Is the implementation example i've put in the comments correct?

@SkafteNicki mind advice, or ping him on slack :]

SkafteNicki commented on docs/source/references/modules.rst:
```diff
@@ -214,6 +214,12 @@ ConfusionMatrix
     :noindex:
 
+CosineSimilarity
+~~~~~~~~~~~~~~~
```
```suggestion
~~~~~~~~~~~~~~~~
```

SkafteNicki commented on torchmetrics/classification/cosine_similarity.py:
```diff
@@ -0,0 +1,91 @@
+class CosineSimilarity(Metric):
+    r"""
+       Computes the `Cosine Similarity <https://en.wikipedia.org/wiki/Cosine_similarity>`_
+        between targets and predictions:
```
just some math :]
```suggestion
        between targets and predictions:
        
        .. math::
            cos_{sim}(x,y) = \frac{x \cdot y}{||x|| \cdot ||y|| = \frac{\sum_{i=1}^n x_i y_i}{\sqrt{\sum_{i=1}^n x_i^2} \sqrt{\sum_{i=1}^n y_i^2}}
            
        where :math:`y` is a tensor of target values, and :math:`x` is a tensor of predictions.
```

SkafteNicki commented on torchmetrics/classification/cosine_similarity.py:
```diff
@@ -0,0 +1,91 @@
+    r"""
+       Computes the `Cosine Similarity <https://en.wikipedia.org/wiki/Cosine_similarity>`_
+        between targets and predictions:
+       Accepts all input types listed in :ref:`references/modules:input types`.
```
This is not correct as it refers to the input types for classification problems.
Instead is should be something similar to 
https://github.com/PyTorchLightning/metrics/blob/652d3f804cad08660842fad440cdead2e986059d/torchmetrics/regression/pearson.py#L36-L39

SkafteNicki commented on torchmetrics/functional/classification/cosine_similarity.py:
```diff
@@ -0,0 +1,63 @@
+) -> Tuple[Tensor, int]:
+    preds.float()
+    target.float()
+
```
call this function here to check that they are same shape
https://github.com/PyTorchLightning/metrics/blob/21ce4f64c9446fe5b98d96ea1dab65caa81441ea/torchmetrics/utilities/checks.py#L23-L26

SkafteNicki commented on torchmetrics/classification/cosine_similarity.py:
```diff
@@ -0,0 +1,91 @@
+           >>> cosine_similarity = CosineSimilarity()
+           >>> cosine_similarity.update(preds, target)
+           >>> cosine_similarity.compute()
+           tensor([1.0000, 1.0000, 1.0000])
```
i am pretty sure the output should be a scalar

SkafteNicki reviewed (commented):
Hi @GiannisVagionakis,
Great work! In my opinion the cosine similarity is not a classification metric but a regression metrics (because the target tensor can also be float values). Therefore please move the files to the appropriate folders.

GiannisVagionakis commented:
> Hi @GiannisVagionakis,
> Great work! In my opinion the cosine similarity is not a classification metric but a regression metrics (because the target tensor can also be float values). Therefore please move the files to the appropriate folders. 

I moved the files to the right folders, and changed the docs too

> i am pretty sure the output should be a scalar

i was actually trying to refer to the case where the reduction arg equals "none", returning the individual scores. 
I changed the docstrings in both files according to yous suggestions

SkafteNicki commented on torchmetrics/functional/regression/cosine_similarity.py:
```diff
@@ -0,0 +1,68 @@
+    target: Tensor,
+) -> Tuple[Tensor, Tensor]:
+    _check_same_shape(preds, target)
+    preds = Tensor.float(preds)
+    target = Tensor.float(target)
```
just preference :]
```suggestion
    preds = preds.float()
    target = target.float()
```

SkafteNicki commented on torchmetrics/functional/regression/cosine_similarity.py:
```diff
@@ -0,0 +1,68 @@
+            >>> from torchmetrics.functional.regression import cosine_similarity
+            >>> target = torch.tensor([[1, 2, 3, 4],
+            ...                        [1, 2, 3, 4]])
+            >>> preds = torch.tensor([[1, 2, 3, 4],
+            ...                       [1, 2, 3, 4]])
+            >>> cosine_similarity(preds, target, 'none')
+            tensor([1.0000, 1.0000])
```
just to not have the precise example twice :]
```suggestion
            >>> preds = torch.tensor([[1, 2, 3, 4],
            ...                       [-1, -2, -3, -4]])
            >>> cosine_similarity(preds, target, 'none')
            tensor([1.0000, -1.0000])
```

SkafteNicki commented on torchmetrics/functional/regression/cosine_similarity.py:
```diff
@@ -0,0 +1,68 @@
+    r"""
+        Computes the `Cosine Similarity <https://en.wikipedia.org/wiki/Cosine_similarity>`_
+        between targets and predictions:
+        Forward accepts
+
+        - ``preds`` (float tensor): ``(N,)``
+        - ``target``(float tensor): ``(N,)``
+
+        Args:
+            preds: Predictions from model (probabilities, logits or labels)
+            target: Ground truth
```
functional has no forward :]
```suggestion

        Args:
            preds: Predicted tensor with shape ``(N,d)``
            target: Ground truth tensor with shape ``(N,d)``
```

SkafteNicki commented on torchmetrics/regression/cosine_similarity.py:
```diff
@@ -0,0 +1,106 @@
+       Computes the `Cosine Similarity <https://en.wikipedia.org/wiki/Cosine_similarity>`_
+        between targets and predictions:
+
+        .. math::
```
could you copy the math to the functional docstring

SkafteNicki commented on torchmetrics/regression/cosine_similarity.py:
```diff
@@ -0,0 +1,106 @@
+
+        Forward accepts
+
+        - ``preds`` (float tensor): ``(N,)``
+        - ``target``(float tensor): ``(N,)``
```
```suggestion
        - ``preds`` (float tensor): ``(N,d)``
        - ``target``(float tensor): ``(N,d)``
```

SkafteNicki commented on torchmetrics/regression/cosine_similarity.py:
```diff
@@ -0,0 +1,106 @@
+        self.reduction = reduction
+
+    def update(self, preds: Tensor, target: Tensor):
+        """
+        Update state with predictions and targets. Forward accepts
+
+        - ``preds`` (float tensor): ``(N,)``
+        - ``target``(float tensor): ``(N,)``
+
+        Args:
+            preds: Predictions from model (probabilities, logits or labels)
+            target: Ground truth labels
+        """
```
```suggestion
        """
        Update metric states with predictions and targets.
        Args:
            preds: Predicted tensor with shape ``(N,d)``
            target: Ground truth tensor with shape ``(N,d)``
        """
```

GiannisVagionakis replied:
@SkafteNicki alright, working on it! Is there anything more needed to be done? for example I saw a failed check on tests with ddp, which I could not run locally as I'm working on a Windows machine

SkafteNicki reviewed (commented):
Great job. A bit a comments to the docstring of metric :]

GiannisVagionakis commented:
@SkafteNicki alright, working on it! Is there anything more needed to be done? for example I saw a failed check on tests with ddp, which I could not run locally as I'm working on a Windows machine (same applies to make-docs)

SkafteNicki commented:
@GiannisVagionakis I will look into the failing test soon :]

SkafteNicki commented:
Hi @GiannisVagionakis,
So I (hopefully) fixed the doctest and have currently removed `none` case from testing as it is making the other test fails.
This has nothing to do with the calculation not being correctly for `none` but has something to do with the underlying testing interface giving data to the comparing function in a different order than your implementation.
I send a follow up PR that corrects this and re-adds the testing of the `none` case.

GiannisVagionakis commented:
@SkafteNicki  @Borda thanks for all the helpfunessl, patience and approachability! Cheers! :]
